### ISSUE TITLE
Crash report

### ISSUE THREAD

**Issue by arnoulfos** [koreader/koreader#10501]
* KOReader version: 2023.04
* Device: hisense a7, Android 10

#### Issue
Crash

#### Steps to reproduce
I toggled reflow on, changed the size of the text to 1.3 and advanced one page

##### `crash.log` (if applicable)
`crash.log` is a file that is automatically created when KOReader crashes. It can normally be found in the KOReader directory:

* `/mnt/private/koreader` for Cervantes
* `koreader/` directory for Kindle
* `.adds/koreader/` directory for Kobo
* `applications/koreader/` directory for Pocketbook

Android logs are kept in memory. Please go to [Menu] → Help → Bug Report to save these logs to a file.

Please try to include the relevant sections in your issue description.
You can upload the whole `crash.log` file (zipped if necessary) on GitHub by dragging and dropping it onto this textbox.

If your issue doesn't directly concern a Lua crash, we'll quite likely need you to reproduce the issue with *verbose* debug logging enabled before providing the logs to us.
To do so, from the file manager, go to [Tools] → More tools → Developer options, and tick both `Enable debug logging` and `Enable verbose debug logging`.
You'll need to restart KOReader after toggling these on.

If you instead opt to inline it, please do so behind a spoiler tag:
<details>
  <summary>crash.log</summary>

```
--------- beginning of system
05-29 09:18:34.936 I/ActivityThread(28799): smart_link is  not supported!!!
05-29 09:18:34.988 D/ActivityThread(28799): add decorView
05-29 09:18:35.000 D/ActivityThread(28799): sdxz decorview lw:-1 lh:-1
05-29 09:18:35.056 I/ActivityThread(28799): refreshActivityViews activity:ActivityRecord{b21ae6f token=android.os.BinderProxy@f6682a6 {org.koreader.launcher.fdroid/org.koreader.launcher.MainActivity}}
```

</details>

**Comment by NiLuJe** [comment]
<need more info> because that log snippet is completely useless, FWIW.

That may be because of a funky device, or not, I dunno, Android's being Android.

In any case, we'll need an actual verbose debug log of this happening.

Random Q: Was the document CJK?

**Comment by hius07** [comment]
Looks similar to https://github.com/koreader/koreader/issues/10150.

**Comment by arnoulfos** [comment]
How can I get the verbose debug log?

**Comment by NiLuJe** [comment]
> If your issue doesn't directly concern a Lua crash, we'll quite likely need you to reproduce the issue with _verbose_ debug logging enabled before providing the logs to us.
> To do so, from the file manager, go to [Tools] → More tools → Developer options, and tick both `Enable debug logging` and `Enable verbose debug logging`.
> You'll need to restart KOReader after toggling these on.

**Comment by arnoulfos** [comment]
After doing what you describe here is the crash log I am getting, strangely it is shorter:
Ljava.lang.Object;@5dea6ee

**Comment by arnoulfos** [comment]
I got it!  here is the log!
--------- beginning of system
05-29 21:22:52.391 D/ActivityThread(17353): sdxz decorview lw:-1 lh:-1
05-29 21:22:52.429 I/ActivityThread(17353): refreshActivityViews activity:ActivityRecord{b38a1b4 token=android.os.BinderProxy@f6682a6 {org.koreader.launcher.fdroid/org.koreader.launcher.CrashReportActivity}}
05-29 21:23:20.740 D/ActivityThread(17353): add decorView
05-29 21:23:20.749 D/ActivityThread(17353): sdxz decorview lw:-1 lh:-1
05-29 21:23:20.799 I/ActivityThread(17353): refreshActivityViews activity:ActivityRecord{9d4455d token=android.os.BinderProxy@4f7c9b {org.koreader.launcher.fdroid/org.koreader.launcher.MainActivity}}
05-29 21:31:21.306 D/ActivityThread(17353): sdxz decorview lw:-1 lh:-1
05-29 21:31:21.333 I/ActivityThread(17353): refreshActivityViews activity:ActivityRecord{9d4455d token=android.os.BinderProxy@4f7c9b {org.koreader.launcher.fdroid/org.koreader.launcher.MainActivity}}
05-29 21:31:22.119 D/ActivityThread(17353): sdxz decorview lw:-1 lh:-1
05-29 21:31:22.130 I/ActivityThread(17353): refreshActivityViews activity:ActivityRecord{9d4455d token=android.os.BinderProxy@4f7c9b {org.koreader.launcher.fdroid/org.koreader.launcher.MainActivity}}
05-29 21:31:37.704 I/ActivityThread(19029): smart_link is  not supported!!!
05-29 21:31:37.767 D/ActivityThread(19029): add decorView
05-29 21:31:37.774 D/ActivityThread(19029): sdxz decorview lw:-1 lh:-1
05-29 21:31:56.745 I/ActivityThread(19029): refreshActivityViews activity:ActivityRecord{d406ec6 token=android.os.BinderProxy@f6682a6 {org.koreader.launcher.fdroid/org.koreader.launcher.CrashReportActivity}}
05-29 21:31:56.792 D/ActivityThread(19029): sdxz decorview lw:-1 lh:-1
05-29 21:31:56.806 I/ActivityThread(19029): refreshActivityViews activity:ActivityRecord{d406ec6 token=android.os.BinderProxy@f6682a6 {org.koreader.launcher.fdroid/org.koreader.launcher.CrashReportActivity}}
05-29 21:31:57.659 D/ActivityThread(19029): sdxz decorview lw:-1 lh:-1
05-29 21:31:57.664 I/ActivityThread(19029): refreshActivityViews activity:ActivityRecord{d406ec6 token=android.os.BinderProxy@f6682a6 {org.koreader.launcher.fdroid/org.koreader.launcher.CrashReportActivity}}
05-29 21:32:44.176 D/ActivityThread(19029): add decorView
05-29 21:32:44.200 D/ActivityThread(19029): sdxz decorview lw:-1 lh:-1
05-29 21:32:44.227 I/ActivityThread(19029): refreshActivityViews activity:ActivityRecord{9d4455d token=android.os.BinderProxy@4f7c9b {org.koreader.launcher.fdroid/org.koreader.launcher.MainActivity}}

**Comment by NiLuJe** [comment]
That's still not relevant, but your Android implementation may be screwing with you there, I dunno, I've always gone through adb myself.

**Comment by arnoulfos** [comment]
Is there something more that i can do?

**Comment by NiLuJe** [comment]
Depends on whether you can get actual logs from something like adb or pidcat ;).

**Comment by ws2bhhh** [comment]
> Depends on whether you can get actual logs from something like adb or pidcat ;).

Would this help?
KOReader version: 2023.05, arm64, from github release.
[crash.txt](https://github.com/koreader/koreader/files/11835879/crash.txt)
The pdf I used for testing is the RP2040 datasheet. When I click on the reflow option, the app crashes immediately.
Also tried the 2022.11 version of koreader and the same problem occurred.

**Comment by ws2bhhh** [comment]
After open the RP2040 datasheet pdf, and then click on the reflow:
06-23 00:00:13.698 14693 14769 I chatty  : uid=10195(org.koreader.launcher) Thread-164 identical 3 lines
06-23 00:00:19.905  4499  4499 D NavigationGestureSideBack: NavigationGestureEdgeBackGestureHandlerHmct, getTopActivity(try): cn = ComponentInfo{org.koreader.launcher/org.koreader.launcher.MainActivity}
06-23 00:00:20.077  4201  1402 D DAAms   : tzyl sat startActivity  original intent = Intent { flg=0x50804000 cmp=org.koreader.launcher/.CrashReportActivity (has extras) } callingPackage=org.koreader.launcher userId=0 inTask null
06-23 00:00:20.077  4201  1402 I ActivityTaskManager: START u0 {flg=0x50804000 cmp=org.koreader.launcher/.CrashReportActivity (has extras)} from uid 10195, pid 14693
06-23 00:00:20.078 14102 14102 D IceBoxCheckService: activityStarting: org.koreader.launcher
06-23 00:00:20.080  4201  1402 D AppLockService: hasAppLock, packageName=org.koreader.launcher,userId=0,res=false
06-23 00:00:20.084 14102 14102 D IceBoxCheckService: activityStarting dbItem: [com.hmct.appicebox.AppUseData@ae07eaa] pkg org.koreader.launcher
06-23 00:00:20.084 14102 14102 D IceBoxCheckService: now): 2023-06-23 00:00:20 pkg org.koreader.launcher dbItem.get(0).lastTimeLong:2023-06-22 23:59:52
06-23 00:00:20.086 14102 14102 D IceBoxCheckService: update: packageNameorg.koreader.launcher
06-23 00:00:20.087  4201  1402 I HmctActivityMonitor: [Activity resume] org.koreader.launcher.CrashReportActivity(org.koreader.launcher)
06-23 00:00:20.090  4201  4399 D SectionHabit: updateFgTimeInner pkgName:org.koreader.launcher,type:5,data:package:org.koreader.launcher ,count :26,launchCountIndex:2,fgtime:41300,fgtimeIndex:0
06-23 00:00:20.090  4201  1402 I ActivityTaskManager: ->startActivity for ActivityRecord{4767310 u0 org.koreader.launcher/.CrashReportActivity t6221} result:START_SUCCESS
06-23 00:00:20.093  4201  4230 D AppLockService: hasAppLock, packageName=org.koreader.launcher,userId=0,res=false
06-23 00:00:20.103  4201  4231 D WindowManagerImpl: sdxz addView pkg:org.koreader.launcher
06-23 00:00:20.105  4201  4231 D WindowManager:  wms-add win to mWindowMap win = Window{61bcad4 u0 Splash Screen org.koreader.launcher},mWindowMap = {android.os.BinderProxy@b1ffd9a=Window{1fdbea8 u0 StatusBar}, android.os.BinderProxy@8fb9e73=Window{d5bb130 u0 com.android.settings/com.android.settings.homepage.SettingsHomepageActivity}, android.os.BinderProxy@e036b49=Window{59b1c4e u0 com.android.systemui.ImageWallpaper}, android.os.BinderProxy@f354d7e=Window{712ab2c u0 AssistPreviewPanel}, android.os.BinderProxy@7e3796d=Window{22819a2 u0 com.android.settings/com.android.settings.SubSettings}, android.os.BinderProxy@8e1b173=Window{df30ca9 u0 NavigationGestureEdgeBackGestureHandlerHmct0}, android.os.BinderProxy@ec1435e=Window{384833f u0 com.hmct.recents/com.hmct.quickstep.RecentsActivity}, android.os.BinderProxy@20e122e=Window{fad375c u0 NavigationGestureSideBack_bottom}, android.os.BinderProxy@c6ca65=Window{90fdbeb u0 NavigationGestureSideBackBottomIndicators}, android.os.BinderProxy@3a5fed7=Window{4cf2ad u0 DockedStackDivider}, android.view.ViewRootImpl$W@3d34827=Window{61bcad4 u0 Splash Screen org.koreader.launcher}, android.os.BinderProxy@2c87b3e=Window{4dd8d9f u0 org.koreader.launcher/org.koreader.launcher.MainActivity}, android.os.BinderProxy@3710b08=Window{8ae52a1 u0 com.hmct.vision/com.hmct.vision.Launcher}, android.os.BinderProxy@92ee8d3=Window{a1b9c10 u0 pl.solidexplorer2/pl.solidexplorer.SolidExplorer}}
06-23 00:00:20.174  4201  4396 W InputDispatcher: channel '4dd8d9f org.koreader.launcher/org.koreader.launcher.MainActivity (server)' ~ Consumer closed input channel or an error occurred.  events=0x9
06-23 00:00:20.174  4201  4396 E InputDispatcher: channel '4dd8d9f org.koreader.launcher/org.koreader.launcher.MainActivity (server)' ~ Channel is unrecoverably broken and will be disposed!
06-23 00:00:20.176  4201  1402 I ActivityManager: Process org.koreader.launcher (pid 14693) has died: fore TOP
06-23 00:00:20.176  4201  1402 V ActivityManager: Clean up application record for ProcessRecord{d4c040 14693:org.koreader.launcher/u0a195} restarting=false allowRestart=true index=-1 replacingPid=false
06-23 00:00:20.176  4201  5160 I WindowManager: WIN DEATH: Window{4dd8d9f u0 org.koreader.launcher/org.koreader.launcher.MainActivity}
06-23 00:00:20.176  4201  4650 E InputDispatcher: Window handle Window{4dd8d9f u0 org.koreader.launcher/org.koreader.launcher.MainActivity} has no registered input channel
06-23 00:00:20.177  4201  5160 W InputDispatcher: Attempted to unregister already unregistered input channel '4dd8d9f org.koreader.launcher/org.koreader.launcher.MainActivity (server)'
06-23 00:00:20.194  4201  1402 I HmctActivityMonitor: [Activity resume(ignore)] org.koreader.launcher.CrashReportActivity(org.koreader.launcher)
06-23 00:00:20.210  4201  4230 D AppLockService: hasAppLock, packageName=org.koreader.launcher,userId=0,res=false
06-23 00:00:20.229  4201  4237 I ActivityManager: Start proc 14815:org.koreader.launcher/u0a195 for activity {org.koreader.launcher/org.koreader.launcher.CrashReportActivity}
06-23 00:00:20.257  4201  4650 W PayGuard: topComp=ComponentInfo{org.koreader.launcher/org.koreader.launcher.CrashReportActivity}packagename=org.koreader.launcher
06-23 00:00:20.257  4201  4650 V WindowManager: Set focused app to: AppWindowToken{9c8e80e token=Token{bde0e09 ActivityRecord{4767310 u0 org.koreader.launcher/.CrashReportActivity t6221}}} moveFocusNow=true displayId=0
06-23 00:00:20.260  4201  4399 D SectionHabit: update pkgName:org.koreader.launcher,type:5,data:package:org.koreader.launcher ,count :26,launchCountIndex:2,fgtime:27494,fgtimeIndex:0
06-23 00:00:20.342 14815 14815 I LockedReading: register receiver action CHANGE_TO_LOCKED_READING for:org.koreader.launcher.CrashReportActivity@ffaca8f
06-23 00:00:20.389 14815 14815 D hmct_activity: onResume pkgName:org.koreader.launcher
06-23 00:00:20.389 14815 14815 D WindowManagerImpl: sdxz addView pkg:org.koreader.launcher
06-23 00:00:20.400  4201  4650 D WindowManager:  wms-add win to mWindowMap win = Window{d2fb082 u0 org.koreader.launcher/org.koreader.launcher.CrashReportActivity},mWindowMap = {android.os.BinderProxy@b1ffd9a=Window{1fdbea8 u0 StatusBar}, android.os.BinderProxy@8fb9e73=Window{d5bb130 u0 com.android.settings/com.android.settings.homepage.SettingsHomepageActivity}, android.os.BinderProxy@e036b49=Window{59b1c4e u0 com.android.systemui.ImageWallpaper}, android.os.BinderProxy@f354d7e=Window{712ab2c u0 AssistPreviewPanel}, android.os.BinderProxy@1419fcd=Window{d2fb082 u0 org.koreader.launcher/org.koreader.launcher.CrashReportActivity}, android.os.BinderProxy@7e3796d=Window{22819a2 u0 com.android.settings/com.android.settings.SubSettings}, android.os.BinderProxy@8e1b173=Window{df30ca9 u0 NavigationGestureEdgeBackGestureHandlerHmct0}, android.os.BinderProxy@ec1435e=Window{384833f u0 com.hmct.recents/com.hmct.quickstep.RecentsActivity}, android.os.BinderProxy@20e122e=Window{fad375c u0 NavigationGestureSideBack_bottom}, android.os.BinderProxy@c6ca65=Window{90fdbeb u0 NavigationGestureSideBackBottomIndicators}, android.os.BinderProxy@3a5fed7=Window{4cf2ad u0 DockedStackDivider}, android.view.ViewRootImpl$W@3d34827=Window{61bcad4 u0 Splash Screen org.koreader.launcher}, android.os.BinderProxy@3710b08=Window{8ae52a1 u0 com.hmct.vision/com.hmct.vision.Launcher}, android.os.BinderProxy@92ee8d3=Window{a1b9c10 u0 pl.solidexplorer2/pl.solidexplorer.SolidExplorer}}
06-23 00:00:20.407  4201  4650 D EpdManagerService: [Activity resume] org.koreader.launcher.CrashReportActivity(org.koreader.launcher)
06-23 00:00:20.446  4201  4650 D WindowManager: Input focus has changed to Window{d2fb082 u0 org.koreader.launcher/org.koreader.launcher.CrashReportActivity}
06-23 00:00:20.523  4201  4231 D WindowManagerImpl: sdxz removeView pkg:org.koreader.launcher
06-23 00:00:20.526  4201 31390 W MydebugSafeInput: -----attribute.packageName = org.koreader.launcher

**Comment by ws2bhhh** [comment]
pidcat version:
[crash_pidcat.log](https://github.com/koreader/koreader/files/11837374/crash_pidcat.log)

**Comment by NiLuJe** [comment]
> arm64

Then it's quite likely the usual old toolchain issue. If your device still supports 32-bit, try the arm build instead?

**Comment by pazos** [comment]
Also `NativeThread  E  Failed to run script: frontend/ui/widget/configdialog.lua:358: attempt to index a number value`

**Comment by NiLuJe** [comment]
Oops, yeah, hadn't checked the pidcat one...

https://github.com/koreader/koreader/blob/4a775b03f1c1b09525808e0ea1660d3dbb7985c6/frontend/ui/widget/configdialog.lua#L357-L361

That one's... weird, though, it's behind a typecheck for `val`, so that one can't be the culprit, which means that the configurable in question itself is the one that's not a table?

**Comment by NiLuJe** [comment]
(I can't repro on a Kobo, FWIW).

**Comment by ws2bhhh** [comment]
> > arm64
> 
> Then it's quite likely the usual old toolchain issue. If your device still supports 32-bit, try the arm build instead?

I tried it. Then the same thing happened and the app crashed immediately.
I'll upload a crash report of the arm build later if I can.

**Comment by NiLuJe** [comment]
Since I can't repro, I'd ideally would have liked more debugging around there (e.g., dumping `self.options[c].values`), unless someone more familiar with ConfigDialog has an idea (@hius07, maybe?).

But, of course, Android makes that very annoying to deal with; unless your device is rooted and you have adb access, @ws2bhhh ?

**Comment by ws2bhhh** [comment]
> Since I can't repro, I would ideally have liked more debugging around there (e.g., dumping `self.options[c].values`), unless someone more familiar with ConfigDialog has an idea (@hius07, maybe?).
> 
> But, of course, Android makes that very annoying to deal with; unless your device is rooted and you have adb access, @ws2bhhh ?

It doesn't seem possible 😔. Hisense A7 differs from the A5 and A9 models in that it does not use a Qualcomm processor but one from UNISOC. There is little information on how to root it.

**Comment by ws2bhhh** [comment]
I'm using version 2024.03, and in the beginning the problem still appeared and pidcat still pointed to a crash at ...configdialog.lua:358... until I (randomly) tried the following actions in roughly the following order:
1. Delete all *.sdr directories
2. Clear all KOReader's user data in Android settings
3. Clear caches in KOReader's settings
4. Disable CRE call cache
5. Opened the RP2040 datasheet and clicked the File browser button **then KOReader immediately crashed** (I accidentally unplugged the cable so I don't know what caused this crash)
6. Reopen KOReader
7. Enable CRE call cache (is this option on by default?)
8. Open the RP2040 datasheet, now reflow works fine.
---
As well as regarding the crash mentioned above, it's very hard to trigger so it shouldn't be a big deal. I can't say for sure if it's the same issue, but I tried many times following the same actions to trigger a similar crash (I tried too many times so I can't upload the full pidcat log):
```
  NativeThread  E  Failed to run script: frontend/document/doccache.lua:72: attempt to compare two nil values
    NativeGlue  V  Pause: 0x740b0bd800

                  Process org.koreader.launcher created for activity {org.koreader.launcher/org.koreader.launcher.CrashReportActivity}
                  PID: 12394   UID:    GIDs:

reader.launche  E  Not starting debugger since process cannot load the jdwp agent.
      MultiDex  I  VM with version 2.1.0 has multidex support
```
Is it worth opening another issues?
